File: 2225.find-players-with-zero-or-one-losses/solution.py
class Solution:
    def findWinners(self, matches: List[List[int]]) -> List[List[int]]:
        wins = defaultdict(int)
        losses = defaultdict(int)
        players = set()
        for winner, loser in matches:
            wins[winner] += 1
            losses[loser] += 1
            players.add(winner)
            players.add(loser)
        
        notLostAnyMatches = []
        loseOnlyOnce = []
        for player in players:
            winCnt = wins[player]
            loseCnt = losses[player]
            if (winCnt > 0) and (loseCnt == 0):
                notLostAnyMatches.append(player)
                continue
            if (loseCnt == 1):
                loseOnlyOnce.append(player)
        return [
            sorted(notLostAnyMatches),
            sorted(loseOnlyOnce)
        ]

File: 2225.find-players-with-zero-or-one-losses/solution.ts
function findWinners(matches: number[][]): number[][] {
  const numsOfLostMap = new Map<number, number>();
  
  const players = new Set<number>();
  for (const [winner, loser] of matches) {
      players.add(winner);
      players.add(loser);
      numsOfLostMap.set(loser, (numsOfLostMap.get(loser) || 0) + 1);
  }
  
  const allWinPlayers: number[] = [];
  const lostOneMatchPlayers: number[] = [];
  for (const player of players.values()) {
      if (!numsOfLostMap.has(player)) {
          allWinPlayers.push(player);
          continue;
      }
      if (numsOfLostMap.get(player) === 1) {
          lostOneMatchPlayers.push(player);
      }
  }
  allWinPlayers.sort((a, b) => a - b);
  lostOneMatchPlayers.sort((a, b) => a - b);
  return [allWinPlayers, lostOneMatchPlayers];
};
